### PR TITLE
Allow organization admins to view pending and rejected invitations

### PR DESCRIPTION
**Pull request by kathy-t**
**Description**
The `OrganizationUser` class currently has the `accepted` boolean, which tracks if the invite is accepted or pending, but it doesn't track if the invite was rejected. This PR adds the `status` field to `OrganizationUser` so that we can track the status of an organization invite and removes the `accepted` field.

There is a UI PR https://github.com/dockstore/dockstore-ui2/pull/1648 that relies on this PR. There is also a compose_setup PR that adds 1.14.0 migrations https://github.com/dockstore/compose_setup/pull/242.

**Review Instructions**
Follow the review instructions in the UI PR https://github.com/dockstore/dockstore-ui2/pull/1648

**Issue**
#4226 
[DOCK-1781](https://ucsc-cgl.atlassian.net/browse/DOCK-1781)

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.

### CODE REVIEW

File: dockstore-webservice/src/main/resources/migrations.1.14.0.xml
Comment (coverbeck) [inline dockstore-webservice/src/main/resources/migrations.1.14.0.xml:32]: Once the migration is done, this will break old web service code that is still running. We often do have downtime with major releases, but we **could** try as an experiment to not have the downtime. So either

1. Can you please note this in the 1.14 deploy notes -- we always hypothesize if the web service will be broken by a release, we should start tracking it when we know it will.
2. Optionally, you could do something with a [trigger](https://stackoverflow.com/a/40901037), that keeps the new column in sync with the old column. Then in 1.15 we drop this column. Probably overkill, but OTOH if we want to eventually get to fully automated deploys, we may need to start doing this.

File: dockstore-integration-testing/src/test/java/io/dockstore/common/CommonTestUtilities.java
Comment (coverbeck) [inline dockstore-integration-testing/src/test/java/io/dockstore/common/CommonTestUtilities.java:58]: 👍 Good fix for the duplication.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java:319]: I didn't try running this, so I might be wrong, but won't this allow all Dockstore users to view all invited/rejected users of an org?

If so, I don't think that should be allowed; only org admins should have access to invites and rejections.

**Review by coverbeck — COMMENTED** [review]
On the no downtime comment, that's more aspirational and Denis and your call if we want to start down that route.

File: dockstore-webservice/src/main/resources/migrations.1.14.0.xml
Comment (denis-yuen) [reply dockstore-webservice/src/main/resources/migrations.1.14.0.xml:32]: I'm ok with either but second that this should either be clearly noted in the deploy notes or deprecated in the second case.

**Review by denis-yuen — APPROVED** [review]

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java
Comment (svonworl) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java:722]: Curious about the difference between the behavior here and in `updateUserRole` around line 756.  If the existing status is REJECTED, this code changes the status to PENDING, but the code in `updateUserRole` throws a bad request.  Why the difference?

Comment (kathy-t) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java:722]: `addUserToOrg` handles rejected invitations by changing the status from `REJECTED` to `PENDING` to "re-add" the user to the organization and the invited user will see the invitation again under their Requests page.

For `updateUserRole`, there isn't much functionality to editing the role if the user has rejected the invitation because it can't be acted upon by the invited user (they can't view their own rejected invites), so I chose to throw a bad request

**Comment by kathy-t** [comment]
Re-requesting reviews for these new changes:
- Brought back the `accepted` field and created a DB trigger that keeps the old `accepted` and new `status` fields in sync. This should allow us to do a no downtime deploy. The old `accepted` field is deprecated. Upon approval, will create a follow-up ticket to remove this deprecated field in 1.15
- Only organization admins can view all memberships. Unauthorized/public users can only view accepted members
- In implementing the previous point, I noticed that we were only checking if a user was an admin/curator without checking that they accepted the invitation so a pending admin/curator can make changes to the organization. I went through all of the endpoints and made sure that we also check that they have accepted the invitation
  - This is more so important with the rejected status because we don't want rejected members to perform admin/maintainer actions

File: dockstore-webservice/src/main/resources/migrations.1.14.0.xml
Comment (kathy-t) [reply dockstore-webservice/src/main/resources/migrations.1.14.0.xml]: Ended up going with number 2, maybe overkill, but it will be interesting to experiment with this

**Review by denis-yuen — APPROVED** [review]

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java:455]: Should maintainers be able to delete their own org? I would argue that they shouldn't, if they're not allowed to update memberships.

This is pretty edge-casey, as it's unlikely that a deleted or rejected org will have any maintainers to begin with.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java:323]: According to the [docs](https://docs.dockstore.org/en/stable/advanced-topics/organizations-and-collections.html?highlight=organizations), only admins can update membership.

If that's the case, should maintainers be able to view rejections/pending? I'd lean towards not, but I don't feel too strongly about it, just raising the question.

File: dockstore-webservice/src/main/resources/migrations.1.14.0.xml
Comment (coverbeck) [inline dockstore-webservice/src/main/resources/migrations.1.14.0.xml:34]: Yowsa, that was more complex than I thought it would be. I see you have a two-way sync as well. Nice!

**Review by coverbeck — APPROVED** [review]
This is great! Thanks for figuring out this path. We can use it as a template for future migrations.

One question, after the migration has been run, and you bring up a new instance of the **old** web service, will that work? I know existing old running code will work, but would the startup schema check pass or fail?

If it fails, I don't know that there's a lot we can do about it. Your code solves the problem of a no-downtime update, but the next thing to potentially plan for is a rollback (something is so wrong with the new version, that we need to temporarily deploy the old version). We've never needed to do this, and hopefully never will, so this is just a question, no need to change anything, and we already have this issue -- just trying to think ahead a little.

Also, if you agree, could you create a followup ticket with a 1.15 milestone to delete the accepted column/field? I suppose we could live with it forever and it would work, but might as well clean it up.

**Comment by denis-yuen** [comment]
> One question, after the migration has been run, and you bring up a new instance of the **old** web service, will that work? I know existing old running code will work, but would the startup schema check pass or fail?

FYI, had this same thought separately in our meeting. As long as we have https://github.com/dockstore/compose_setup/blob/develop/templates/web.yml.template#L115 existing webservices will continue running if currently running, but will kill themselves on start-up
We probably need a ticket to turn this off in, say 1.14.0, so when we delete the column in 1.14.1, the 1.14.0 service will stillbe able to start if there's an issue

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/OrganizationUser.java
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/OrganizationUser.java:54]: Belated thought after this came up in a different meeting with @denis-yuen  -- do we need this at all?

I guess it depends on how the Hibernate schema validator works. Does it allow extra columns that aren't mapped? According to this unverified [StackOverflow question](https://stackoverflow.com/questions/39493603/hibernate-schema-validation-fail-on-additional-columns-tables) it does. The user is complaining that it does, but that would work to our advantage.

If that's true, then I think you can just remove this from the Java code entirely. The 1.14 code will access and only know about the new column, the 1.13 code will access and only know about the old column, and the DB keeps them in sync behind the scenes. Untested theory.

File: dockstore-integration-testing/src/test/java/io/dockstore/common/CommonTestUtilities.java
Comment (svonworl) [inline dockstore-integration-testing/src/test/java/io/dockstore/common/CommonTestUtilities.java:239]: Entirely out of curiosity, and no need to change this, but could this line do all the work by itself, no `if` necessary?

**Review by svonworl — APPROVED** [review]

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java
Comment (kathy-t) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java:323]: This uses `isUserAdmin` instead of `isUserAdminOrMaintainer` so only organization admins can view rejections/pending.

File: dockstore-integration-testing/src/test/java/io/dockstore/common/CommonTestUtilities.java
Comment (kathy-t) [reply dockstore-integration-testing/src/test/java/io/dockstore/common/CommonTestUtilities.java:239]: Yeah you're right, the `if` isn't necessary

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java
Comment (kathy-t) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java:455]: Will change it back so only the organization admin can delete the org

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/OrganizationUser.java
Comment (kathy-t) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/core/OrganizationUser.java:54]: I tested this by removing the `accepted` field and the webservice starts with no errors, but I believe this will fail the CircleCI build [here](https://github.com/dockstore/dockstore/blob/develop/.circleci/config.yml#L292) because the migrations are not consistent with the JPA class

**Comment by kathy-t** [comment]
> One question, after the migration has been run, and you bring up a new instance of the **old** web service, will that work? I know existing old running code will work, but would the startup schema check pass or fail?
> 
> If it fails, I don't know that there's a lot we can do about it. Your code solves the problem of a no-downtime update, but the next thing to potentially plan for is a rollback (something is so wrong with the new version, that we need to temporarily deploy the old version). We've never needed to do this, and hopefully never will, so this is just a question, no need to change anything, and we already have this issue -- just trying to think ahead a little.

I tried testing this locally by running the 1.14 migrations then starting the develop webservice (aka an old webservice without 1.14 migrations) and the webservice started with no schema errors. Keep in mind that if we actually do rollback in AWS, the 1.13 migrations will run before the old webservice container starts since every task has its own migrations container.

> Also, if you agree, could you create a followup ticket with a 1.15 milestone to delete the accepted column/field? I suppose we could live with it forever and it would work, but might as well clean it up.

Created https://github.com/dockstore/dockstore/issues/5256

**Review by coverbeck — COMMENTED** [review]
Was here, going to wait for further implementation per our Slack huddle.

**Comment by kathy-t** [comment]
My newest change removes the `accepted` field from the `OrganizationUser` class. It **does not** remove the `accepted` column. I had to add a hack in import.sql so that the [CircleCI step](https://github.com/dockstore/dockstore/blob/cefa5db4faafb25d36bb5c0df714ae5a7ce5a7aa/.circleci/config.yml#L292) that checks if JPA classes are consistent with migrations passes. I added a comment in the file.

Pros:
- 1.13 code only knows about `accepted` and 1.14 code only knows about `status`
- After deploying 1.14, we can still rollback to 1.13 if needed because the `accepted` column still exists in the database
- In 1.15, when we drop the `accepted` column and triggers in the database, if we need to roll back to 1.14, the webservice should be able to start without schema validation errors because the class doesn't have the `accepted` field
  - I tested this scenario locally by running migrations that drop the triggers and the `accepted` column, then starting the webservice from this PR and it started successfully.

Table summarizing the state of `accepted` and `status` through the releases:
- field = Java field in the `OrganizationUser` class
- column = database column

| Release | Change                                                                         | `accepted` field   | `accepted` column  | `status` field     | `status` column    |
|---------|--------------------------------------------------------------------------------|--------------------|--------------------|--------------------|--------------------|
| 1.13    | n/a                                                                            | :heavy_check_mark: | :heavy_check_mark: |                    |                    |
| 1.14    | Remove `accepted` field, create `status` field and column, and create triggers |                    | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| 1.15    | Remove `accepted` column and triggers                                          |                    |                    | :heavy_check_mark: | :heavy_check_mark: |

**Review by svonworl — APPROVED** [review]

**Review by coverbeck — APPROVED** [review]
Nice!

**Review by denis-yuen — APPROVED** [review]

**Pull request merged**
